library/core/src/sync/atomic.rs

```diff
@@ -79,6 +79,30 @@
 //!
 //! [lock-free]: https://en.wikipedia.org/wiki/Non-blocking_algorithm
 //!
+//! # Atomic accesses to read-only memory
+//!
+//! In general, atomic accesses on read-only memory are Undefined Behavior. For instance, attempting
+//! to do a `compare_exchange` that will definitely fail (making it conceptually a read-only
+//! operation) can still cause a page fault if the underlying memory page is mapped read-only. Since
+//! atomic `load`s might be implemented using compare-exchange operations, even a `load` can fault
+//! on read-only memory.
+//!
+//! However, as an exception from this general rule, Rust guarantees that "sufficiently small"
+//! atomic loads are implemented in a way that works on read-only memory. This threshold of
+//! "sufficiently small" depends on the architecture:
+//!
+//! | Target architecture | Maximal atomic `load` size that is guaranteed read-only |
+//! |-----------|---------|
+//! | `x86`     | 4 bytes |
+//! | `x86_64`  | 8 bytes |
+//! | `arm`     | 4 bytes |
+//! | `aarch64` | 8 bytes |
+//! | `riscv32` | 4 bytes |
+//! | `riscv64` | 8 bytes |
+//!
```

Comment:
```suggestion
//! | `powerpc64le-unknown-linux-gnu` | 16 bytes |
//!
```

LLVM generates calls to `__atomic_load_16` which always uses the `lq` instruction -- GCC and LLVM disagree on the instruction being enabled by default.

Reply:
Is that really specific to the entire triple, or does this apply to all powerpc64le targets?

Reply:
i'm not 100% sure (since the PowerISA spec is somewhat ambiguous), but it should also apply to aix and linux kernel

Reply:
I'm confused. This is entirely a property of the codegen backend, right? For which sizes of atomic access does LLVM emit operations that work on read-only pages? This shouldn't depend on the OS/vendor part of the triple, right?

Reply:
it does because different OSes supply different libatomic implementations and supply different default cpu feature sets. e.g. MacOS requires x86-64-v2 (approximately)

Reply:
on Power, AIX requires many more cpu features than Linux.

Reply:
Can those features be opted-out via `-C target-cpu/target-feature`? The number we give here is the minimum that we guarantee for all configurations of the target.

Reply:
> Can those features be opted-out via `-C target-cpu/target-feature`?
yes, but all that does is make llvm call `__atomic_load_16` which, for linux, just does the atomic load using the `lq` instruction you just opted out of using. in fact, llvm has that feature disabled by default on linux (gcc has it enabled by default)

Reply:
Okay.

If AIX and maxOS require more than Linux, would it then be correct to list 16 as a lower bound for all powerpc64le targets?

Reply:
> If AIX and maxOS require more than Linux, would it then be correct to list 16 as a lower bound for all powerpc64le targets?

modern macOS doesn't support PowerISA and hasn't for a long time iirc, it was an example for requiring additional features on x86-64.

16 is the correct lower bound for Linux with glibc and (according to the PowerISA spec.) AIX, I'd have to check musl and freebsd.

Reply:
8 is the lower bound when clang is the compiler that libatomic or equivalent was built with, since clang doesn't enable the `lq` instruction by default.